### ISSUE TITLE
LadybugDynamo Memory issues

### ISSUE THREAD

**Issue by mostaphaRoudsari** [ladybug-analysis-tools/ladybug-dynamo#8]
I'm opening this issue to document the issue that I have with memory management in Dynamo.
## Preparation

To replicate this issue you need to [download the zip file](https://github.com/ladybug-analysis-tools/ladybug-dynamo/archive/master.zip) this repository or clone it. Create a Ladybug folder under Dynamo packages folder and copy ladybugdynamo folder under extra (e.g. `C:\Users\Administrator\AppData\Roaming\Dynamo\0.9\packages\Ladybug\extra`) Then copy the files under [source folder](https://github.com/ladybug-analysis-tools/ladybug-dynamo/tree/master/plugin/source) to the same folder. The folder will look like: 

![image](https://cloud.githubusercontent.com/assets/2915573/12346387/d719ced6-bb21-11e5-8457-977e759ba993.png)
## Memory Issues

Next I fire Dynamo - sandbox daily built. Before making any move it takes **155.2 MB** of the memory.

![image](https://cloud.githubusercontent.com/assets/2915573/12346549/575d15ac-bb23-11e5-9496-0345726d0a38.png)

Assuming that you followed the preparation open [sunlight hours sample file](https://github.com/ladybug-analysis-tools/ladybug-dynamo/tree/master/plugin/sampleFiles) under examples folder: `sunlighthoursAnalysis.dyn`. As you can see below once the analysis is finished it takes **2,197 MB** of the memory! What the code does is solving intersection for 10200 \* number of sun vectors \* number of geometries.

![image](https://cloud.githubusercontent.com/assets/2915573/12346603/ebb5e4cc-bb23-11e5-8c86-2d0fb1ccca2c.png)

I create a [LBTestPt](https://github.com/ladybug-analysis-tools/ladybug-dynamo/blob/master/ladybugdynamo/geometryoperations.py#L30-L54) for each pt and run the intersections. I don't collect any of the intersections and write them all to the same variable so I don't think that is the place that causes the issue. After reading [this link](https://github.com/DynamoDS/Dynamo/wiki/Zero-Touch-Plugin-Development#dispose--using-statement) I started to wonder if this is because all the [lineRays](https://github.com/ladybug-analysis-tools/ladybug-dynamo/blob/master/ladybugdynamo/geometryoperations.py#L61-L79) that I'm creating for each point. However this is a big issue let's assume this is an issue with my code which I or someone else will fix eventually and move on to the bigger issue.

Go ahead and change the month from 12 to 6.
![image](https://cloud.githubusercontent.com/assets/2915573/12346812/e3d798a2-bb25-11e5-8805-b37849b12f47.png)

The analysis will re-run and once it's over the memory use is **3,168.5 MB**. If I make any changes at this point and run the analysis again my laptop will run out of memory.

![image](https://cloud.githubusercontent.com/assets/2915573/12346872/523c79ca-bb26-11e5-93a8-ca973ab4a57d.png)

What makes this even more problematic is even after I close the file it doesn't clean the memory until I close the Dynamo itself.

![image](https://cloud.githubusercontent.com/assets/2915573/12346919/9fff2dba-bb26-11e5-991f-9535be66132b.png)

I hope this explains the issue. Let me know if there is anything else that you need me to provide. 

Also in case you don't want to download the repository just download [this file](https://dl.dropboxusercontent.com/u/16228160/Ladybug.zip) and copy the folder under packages folder on your machine.

Finally if you want to test a similar issue in a lower scale with no geometry operation, try importEpw under [sampleFiles](https://github.com/ladybug-analysis-tools/ladybug-dynamo/tree/master/plugin/sampleFiles) and load a number of weather files.

cc: @ikeough, @tatlin, @mjkkirschner, @mccrone

**Comment by mjkkirschner** [comment]
@mostaphaRoudsari I will need some more time to look into the issue further but there are some hints in my initial profiling and looking at your code:

In the images below you can see that there are about 3 gigs of unmanaged objects we cannot inspect using .net profilers. This immediately makes me suspect the geometry operations and of course the fact that there are no disposal calls in your python code.

##### If you do not implement dispose() on your types (and call it and make it dispose references to geometry) or dispose() any intermediate geometry (protogeometry types) before you return your code will cause strange crashes and hangs. It will happen. If you do not return geometry into Dynamo, you must dispose it. **

When you said in your email that disposing the vectors makes no difference - do you mean you see no difference in the task manager? Or that your machine still crashes? It would be best to do a sweep through your entire source and attempt to find all intermediate geometry, and other unmanaged resources and dispose it and see if you still have crashes.

I ask because I do not think you will not see a difference in the task manager. Protogeometry will most likely not return the memory until requested, BUT you should not get a crash if you are disposing everything correctly.

Are there other unmanaged resources that you are not disposing? ie. file handles, images, etc?

There is another issue with the python engines. I believe that each engine that is loaded loads all dependencies again, we do not cache these as the contexts are completely separate and created whenever the python node runs, that means whenever a python script is loaded all the imported modules are loaded again. (including many calls to load DynamoCore, CLR, protogeometry etc)....

Once a dll is loaded I'm not sure it can be unloaded / cleaned with our current Python node setup.

I'll keep looking into what's going on and let you know what I find.

<img width="1008" alt="screen shot 2016-01-19 at 12 12 09 am" src="https://cloud.githubusercontent.com/assets/508936/12427676/93a392aa-beae-11e5-8532-0eaea9aae386.png">
<img width="1913" alt="screen shot 2016-01-19 at 12 03 52 am" src="https://cloud.githubusercontent.com/assets/508936/12427677/93aaeed8-beae-11e5-99d1-eeae00ff674b.png">
<img width="877" alt="screen shot 2016-01-18 at 11 04 01 pm" src="https://cloud.githubusercontent.com/assets/508936/12427678/93ab6962-beae-11e5-811a-237c89630e94.png">

**Comment by mostaphaRoudsari** [comment]
Hi @mjkkirschner, Thank you for checking. I admit that I used the wrong term. My machine doesn't actually crash. It does run out of memory.

I'm double checking the code and make sure to dispose geometry interactions as well as the python objects itself except what I output in Dynamo. I will let you know when I commit the newer version so you can give it another try.

I also started doing comparative studies between Dynamo and Grasshopper for non-geometric libraries. I will add the results here.

**Comment by mjkkirschner** [comment]
@mostaphaRoudsari we should still be able to figure out how to stop you from running out of memory, because you're seeing these issues without geometry problems as well (is that true?) then if forced to guess I would say it has to do with python and the loading of libraries repeatedly.

**Comment by mostaphaRoudsari** [comment]
@mjkkirschner I have a good news and a bad news. The good news is disposing the geometries and also deleting the objects that I create in Python except the results of the component helped with the memory but then it makes the script much slower. I will work on it a little bit more and then I'll commit the final version to repository so you can give it a try tomorrow.

**Comment by pflaumingo** [comment]
@mostaphaRoudsari has this issue been fixed? I'm having some issues on my end with memory after successive runs.

**Comment by mostaphaRoudsari** [comment]
Hi @pflaumingo! This one is not solved but I'm developing a radiance-based method for the new release which will solve the issue for users. I will keep you posted.

**Comment by schaggar16** [comment]
@mostaphaRoudsari, My machines RAM (32GB) keeps getting maxed out when i run solar radiation analysis using Ladybug 0.1.5. I am running Dynamo 1.2.1 and Revit 2016, is there a fix for this?

**Comment by mostaphaRoudsari** [comment]
@schaggar16 and @pflaumingo I replied here: https://github.com/ladybug-tools/ladybug-dynamo/issues/27#issuecomment-304537686

**Comment by mostaphaRoudsari** [comment]
All the analysis are now moved to [honeybee-dynamo](https://github.com/ladybug-tools/honeybee-dynamo) and are using Radiance for ray-tracing. This is not related anymore however it would be great if this will be addressed inside Dynamo at some point.
